codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/UnionGenerator.java

```diff
@@ -51,48 +52,41 @@ public void run() {
             writer.write("is$L()", symbol.getName());
         }).write("");
 
-        // Create type aliases for each member that satisfy the interface.
+        // Create structs for each member that satisfy the interface.
         for (MemberShape member : shape.getAllMembers().values()) {
             Symbol memberSymbol = symbolProvider.toSymbol(member);
             String exportedMemberName = symbol.getName() + symbolProvider.toMemberName(member);
-            String unExportedMemberName = StringUtils.uncapitalize(exportedMemberName);
+            Shape target = model.expectShape(member.getTarget());
 
-            // Create an interface for the member
+            // Create the member's concrete type
             writer.writeMemberDocs(model, member);
-            writer.openBlock("type $L interface {", "}", exportedMemberName, () -> {
-                writer.write("$L", symbol.getName());
-                writer.write("is$L()", exportedMemberName);
-                writer.write("Value() $T", memberSymbol);
+            writer.openBlock("type $L struct {", "}", exportedMemberName, () -> {
+                // Union members can't have null values, so for simple shapes we don't
+                // use pointers. We have to use pointers for complex shapes since,
+                // for example, we could still have a map that's empty or which has
+                // null values.
+                if (target instanceof SimpleShape) {
+                    writer.write("Value $T", memberSymbol);
+                } else {
+                    writer.write("Value $P", memberSymbol);
+                }
```

Comment:
Not specific to this PR, but at some point we should take a look at how slice/map types are generated from `$T`, and `$P`. If a member is or is not a pointer of a slice/map should not be driven by `$T`, or `$P`, the member being a pointer is driven by the modeled list/map shape, and its member.

This mean that `$T` should generate the correct Go type for a map of strings, i.e. `map[string]*string`. In addition the same for lists (assuming lists cannot be sparse, i.e. `[]string`. Using `$P` should generate the pointer type of the outer shape, `*map[string]*string` and `*[]string`.  `$P` should have no impact on the generate of nested members within collections.

Reply:
created, https://github.com/awslabs/smithy-go/issues/149 for this.